skbio/diversity/_block.py

```diff
@@ -0,0 +1,280 @@
+
+
+def _generate_id_blocks(ids, k=64):
+    """Generate blocks of IDs that map into a DistanceMatrix
+
+    Parameters
+    ----------
+    ids : Iterable object
+        An iterable of IDs of whatever type.
+    k : int, optional
+        The size of a block to generate IDs for, defaults to 64.
+
+    Notes
+    -----
+    This method is intended to facilitate partial beta diversity calculations.
+    Blocks of IDs are generated from the upper triangle of the subsequent
+    distance matrix. For instance, given the following distance matrix with
+    IDs {A, B, C, D, E}:
+
+      A B C D E
+    A 0 # # # #
+    B # 0 # # #
+    C # # 0 # #
+    D # # # 0 #
+    E # # # # 0
+
+    The goal of this method is to generate tuples of IDs of at most size k over
+    the upper triangle which correspond to blocks of the matrix to compute. IDs
+    are remapped as well into integers to facilitate downstream indexing.
+    Given k=3, the following ID tuples would be generated:
+
+    ((0, 1, 2), (0, 1, 2))
+    ((0, 1, 2), (3, 4))
+    ((3, 4), (3, 4))
+
+    This method is not responsible for describing which specific pairs of IDs
+    are to be computed, only the subset of the matrix of interest.
+
+    Returns
+    -------
+    tuple of 1D np.array
+        Index 0 contains the row IDs, and index 1 contains the column IDs
+    """
+    n = len(ids)
+    ids_idx = np.arange(n)
+
+    for row_start in range(0, n, k):
+        for col_start in range(row_start, n, k):
+            row_ids = ids_idx[row_start:row_start + k]
+            col_ids = ids_idx[col_start:col_start + k]
+
+            yield (row_ids, col_ids)
+
+
+def _block_party(counts=None, row_ids=None, col_ids=None, **kwargs):
+    """Subset counts to relevant rows and columns
+
+    Parameters
+    ----------
+    counts : 2D array_like of ints or floats
+        Matrix containing count/abundance data where each row contains counts
+        of OTUs in a given sample.
+    row_ids : 1D np.ndarray of int
+        Block row IDs to keep in the counts matrix.
+    col_ids : 1D np.ndarray of int
+        Block column IDs to keep in the counts matrix. Note, these correspond
+        to rows in the counts matrix, but columns in a subsequent distance
+        matrix.
+    kwargs : dict
+        Keyword arguments containing information about the block to compute.
+
+    Returns
+    -------
+    dict
+        kwargs that describe the block to compute. A filtered ``counts`` matrix
+        is stored in kwargs. If applicable, a filtered ``tree`` and ``otu_ids``
+        are also stored.
+    """
+    # filter kwargs to valid keys for downstream block compute
+    # explicit copies are not necessary as implicit copies or overwwrites are
+    # performed below.
+    valid_block_keys = {'counts', 'ids', 'tree', 'otu_ids', 'metric',
+                        'id_pairs', 'validate'}
+    block_kwargs = {k: v for k, v in kwargs.items() if k in valid_block_keys}
```

Comment:
Actually it might be best to pull the validation out entirely as this gets run many times? Maybe `_block_kwargs` could inherit this? I would assume anything returned from `_block_kwargs` could be considered "safe".